core/src/main/scala/chisel3/DataMirror.scala

```diff
@@ -0,0 +1,301 @@
+    }
+  }
+
+  /** Return all expanded components, including intermediate aggregate nodes
+    *
+    * @param d Component to find leafs if aggregate typed. Intermediate fields/indicies ARE included
+    */
+  def getIntermediateAndLeafs(d: Data): Seq[Data] = collectDeep(d) { case x => x }.toSeq
+
+  /** Collects all fields selected by collector within a data and all recursive children fields
+    * Accepts a collector partial function, rather than a collector function
+    *
+    * @param data Data to collect fields, as well as all children datas it directly and indirectly instantiates
+    * @param collector Collector partial function to pick which components to collect
+    * @tparam T Type of the component that will be collected
+    */
+  def collectDeep[T](d: Data)(collector: PartialFunction[Data, T]): Iterable[T] = {
+    val myItems = collector.lift(d)
+    val deepChildrenItems = d match {
+      case a: Aggregate => a.getElements.flatMap { x => collectDeep(x)(collector) }
+      case other => Nil
+    }
+    myItems ++ deepChildrenItems
+  }
+
+  // Alignment-aware collections
+  import connectable.{AlignedWithRoot, Alignment, FlippedWithRoot}
+  // Implement typeclass to enable collecting over Alignment
+  implicit val AlignmentMatchingZipOfChildren = new HasMatchingZipOfChildren[Alignment] {
+    def matchingZipOfChildren(
+      left:  Option[Alignment],
+      right: Option[Alignment]
+    ): Seq[(Option[Alignment], Option[Alignment])] =
+      Alignment.matchingZipOfChildren(left, right)
+  }
+
+  /** Collects all members of base who are aligned w.r.t. base
+    * Accepts a collector partial function, rather than a collector function
+    *
+    * @param base Data from whom aligned members (w.r.t. base) are collected
+    * @param collector Collector partial function to pick which components to collect
+    * @tparam T Type of the component that will be collected
+    */
+  def collectAlignedDeep[T](base: Data)(pf: PartialFunction[Data, T]): Seq[T] = {
+    collectDeepOverAllForAny(Some(Alignment(base, Set.empty, true)), None) {
+      case (Some(x: AlignedWithRoot), _) => (pf.lift(x.member), None)
+    }.map(_._1).flatten
+  }
+
+  /** Collects all members of base who are flipped w.r.t. base
+    * Accepts a collector partial function, rather than a collector function
+    *
+    * @param base Data from whom flipped members (w.r.t. base) are collected
+    * @param collector Collector partial function to pick which components to collect
+    * @tparam T Type of the component that will be collected
+    */
+  def collectFlippedDeep[T](base: Data)(pf: PartialFunction[Data, T]): Seq[T] = {
+    collectDeepOverAllForAny(Some(Alignment(base, Set.empty, true)), None) {
+      case (Some(x: FlippedWithRoot), _) => (pf.lift(x.member), None)
+    }.map(_._1).flatten
+  }
+
+  // Typeclass trait to use collectDeepOverMatches, collectDeepOverAll, collectDeepOverAllForAny, collectDeepOverAllForAnyFunction
```

Comment:
scaladoc? not sure how much of this stuff is meant to be exposed to users

Reply:
It is available to power users, but I'll add scaladoc.